lib/segment/client/http.ex

```diff
@@ -0,0 +1,150 @@
+    ]
+
+    Tesla.client(middleware, adapter)
+  end
+
+  def call(client, events) when is_list(events), do: batch(client, events)
+
+  def call(client, event) do
+    case callp(client, event.type, prepare_events(event), @retry_attempts) do
+      {:ok, %{status: status}} when status == 200 ->
+        :ok
+
+      {:ok, %{status: status}} when status == 400 ->
+        Logger.error("[Segment] Call Failed. JSON too large or invalid")
+        :error
+
+      {:error, err} ->
+        Logger.error("[Segment] Call Failed after #{@retry_attempts} retries. #{inspect(err)}")
+        :error
+
+      err ->
+        Logger.error("[Segment] Call Failed #{inspect(err)}")
+        :error
+    end
+  end
+
+  def batch(client, events, context \\ nil, integrations \\ nil) do
+    data =
+      %{batch: prepare_events(events)}
+      |> add_if(:context, context)
+      |> add_if(:integrations, integrations)
+
+    case callp(client, "batch", data, @retry_attempts) do
+      {:ok, %{status: status}} when status == 200 ->
+        :ok
+
+      {:ok, %{status: status}} when status == 400 ->
+        Logger.error(
+          "[Segment] Batch call of #{length(events)} events failed. JSON too large or invalid"
+        )
+
+        :error
+
+      {:error, err} ->
+        Logger.error(
+          "[Segment] Batch call of #{length(events)} events failed after #{@retry_attempts} retries. #{
+            inspect(err)
+          }"
+        )
+
+        :error
+
+      err ->
+        Logger.error("[Segment] Batch callof #{length(events)} events failed #{inspect(err)}")
+        :error
+    end
+  end
+
+  defp callp(client, url, data, retries) when retries > 0 do
+    retry with: linear_backoff(@retry_start, 2) |> cap(@retry_expiry) |> Stream.take(retries) do
+      Tesla.post(client, url, data)
+    after
+      result -> result
+    else
+      error -> error
+    end
+  end
+
+  defp callp(client, url, data, _retries) do
+    Tesla.post(client, url, data)
+  end
+
+  defp prepare_events(items) when is_list(items), do: Enum.map(items, &prepare_events/1)
+
+  defp prepare_events(item) do
+    Map.from_struct(item)
+    |> prep_context()
+    |> add_timestamp()
+    |> drop_nils()
+  end
+
+  defp drop_nils(map) do
+    map
+    |> Enum.filter(fn
+      {_, %{} = item} when map_size(item) == 0 -> false
+      {_, nil} -> false
+      {_, _} -> true
+    end)
+    |> Enum.into(%{})
+  end
+
+  defp prep_context(%{context: nil} = map),
+    do: %{map | context: map_content(Segment.Analytics.Context.new())}
+
+  defp prep_context(%{context: context} = map), do: %{map | context: map_content(context)}
+
+  defp prep_context(map),
+    do: Map.put_new(map, :context, map_content(Segment.Analytics.Context.new()))
+
+  defp map_content(%Segment.Analytics.Context{} = context), do: Map.from_struct(context)
+  defp map_content(context) when is_map(context), do: context
+
+  defp add_timestamp(%{timestamp: nil} = map), do: Map.put(map, :timestamp, DateTime.utc_now())
```

Comment:
Is this the correct format for the timestamp? If I remember correctly this would make it an ISO-8601 timestamp?

Reply:
yep thats the right format according to Segment docs